File: coursera/kotlin-for-java-developers/week1/src/main/java/extensions/callingextensions/CallingExtensions.java
package extensions.callingextensions;

public class CallingExtensions {

    static class Parent {
        public String memberFoo() {
            return "parent";
        }

        public String foo() {
            return "parent";
        }

        public String foo(String string) {
            return "parent: " + string;
        }
    }

    static class Child extends Parent {
        public String memberFoo() {
            return "child";
        }

        public String foo() {
            return "child";
        }

        public String foo(String string) {
            return "child: " + string;
        }
    }

    public static String foo(Parent parent) {
        return "parent";
    }

    public static String foo(Child child) {
        return "child";
    }

    public static String foo(Parent parent, String string) {
        return "parent: " + string;
    }

    public static String foo(Child child, String string) {
        return "child: " + string;
    }

    public static void main(String[] args) {
        Parent parent = new Child();
        System.out.println(foo(parent));
        System.out.println(parent.memberFoo());
        System.out.println(parent.foo());
        System.out.println(parent.foo("test"));
        System.out.println(foo(parent, "test"));
    }
}

File: coursera/kotlin-for-java-developers/week2/src/main/kotlin/extensions/callingextensions/CallingExtensions.kt
package extensions.callingextensions

open class Parent {
    open fun memberFoo() = "member parent"
    open fun foo(string: String) = "member parent: $string"
}

class Child : Parent() {
    override fun memberFoo() = "member child"
    override fun foo(string: String) = "member child: $string"
}

fun Parent.foo() = "extension parent"
fun Child.foo() = "extension child"

// extension functions shadowed
@Suppress("EXTENSION_SHADOWED_BY_MEMBER")
fun Parent.foo(string: String) = "extension parent: $string"

@Suppress("EXTENSION_SHADOWED_BY_MEMBER")
fun Child.foo(string: String) = "extension child: $string"

// need to change jvm name to avoid bytecode clash with extension functions
@JvmName("fooJvm")
fun foo(parent: Parent): String = "static parent"

@JvmName("fooJvm")
fun foo(child: Child): String = "static child"

fun main() {
    val parent: Parent = Child()
    println(parent.foo())
    println(parent.memberFoo())
    println(parent.foo("test"))
    println(foo(parent))
}